### PR TITLE
Add support for Velocity IP forwarding

### PR DESCRIPTION
**Pull request by astei**
While Velocity already supports BungeeCord-style IP forwarding (which almost everyone supports now), it is not secure. Users have a lot of problems setting up firewalls or setting up plugins like iPWhitelist. Further, the BungeeCord IP forwarding protocol still retains essentially its original form, when there is brand new support for custom login plugin messages in 1.13.

Velocity's modern IP forwarding uses an HMAC-SHA256 code to ensure authenticity of messages (users have to set up an unique secret), is packed into a binary format that is smaller than BungeeCord's forwarding, and is integrated into the Minecraft login process by using the 1.13 login plugin message packet.

I have verified this with Velocity build 303 in online-mode.

### CODE REVIEW

File: Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+     }
++    public UUID readUuid() { return this.i(); } // Paper - OBFHELPER
+ 
+     public PacketDataSerializer d(int i) {
+         while ((i & -128) != 0) {
+@@ -316,6 +318,7 @@ public class PacketDataSerializer extends ByteBuf {
+             }
+         }
+     }
++    public String readUtf(int maxLength) { return this.e(maxLength); } // Paper - OBFHELPER
```
Comment (aikar) [inline Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch]: move down a line, change to UTF

File: Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+     }
++    public int readVarInt() { return this.g(); } // Paper - OBFHELPER
+ 
+     public long h() {
+         long i = 0L;
+@@ -183,6 +184,7 @@ public class PacketDataSerializer extends ByteBuf {
+     public UUID i() {
+         return new UUID(this.readLong(), this.readLong());
+     }
++    public UUID readUuid() { return this.i(); } // Paper - OBFHELPER
```
Comment (aikar) [inline Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch]: move down a line, change to UUID

File: Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+         } else {
+diff --git a/src/main/java/net/minecraft/server/PacketDataSerializer.java b/src/main/java/net/minecraft/server/PacketDataSerializer.java
+index cab837483..c1d5a5668 100644
+--- a/src/main/java/net/minecraft/server/PacketDataSerializer.java
++++ b/src/main/java/net/minecraft/server/PacketDataSerializer.java
+@@ -156,6 +156,7 @@ public class PacketDataSerializer extends ByteBuf {
+ 
+         return i;
+     }
++    public int readVarInt() { return this.g(); } // Paper - OBFHELPER
```
Comment (aikar) [inline Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch]: move down a line

File: Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+ 
+diff --git a/src/main/java/net/minecraft/server/NameReferencingFileConverter.java b/src/main/java/net/minecraft/server/NameReferencingFileConverter.java
+index 8a2ff6a41..d9839cb48 100644
+--- a/src/main/java/net/minecraft/server/NameReferencingFileConverter.java
++++ b/src/main/java/net/minecraft/server/NameReferencingFileConverter.java
+@@ -61,6 +61,7 @@ public class NameReferencingFileConverter {
+         });
+ 
+         if (minecraftserver.getOnlineMode()
++                || (com.destroystokyo.paper.PaperConfig.velocitySupport && com.destroystokyo.paper.PaperConfig.velocityOnlineMode) // Paper - Velocity support
```
Comment (aikar) [inline Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch]: We really need to abstract all this now to a single method PaperConfig.isProxyOnlineMode()

File: Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
++import java.nio.charset.StandardCharsets;
+ import java.util.HashMap;
+ import java.util.List;
+ import java.util.Map;
+@@ -69,7 +70,7 @@ public class PaperConfig {
+         commands.put("paper", new PaperCommand("paper"));
+ 
+         version = getInt("config-version", 14);
+-        set("config-version", 15);
++        set("config-version", 16);
```
Comment (aikar) [inline Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch]: Not needed. Just add getXX calls in the method and itll add the defaults for you.

File: Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+@@ -433,4 +434,26 @@ public class PaperConfig {
+             }
+         }
+     }
++
++    public static boolean velocitySupport;
++    public static boolean velocityOnlineMode;
++    public static byte[] velocitySecretKey;
++    private static void velocitySupport() {
++        if (version < 16) {
```
Comment (aikar) [inline Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch]: as said above, all the below getXX methods will set these defaults for you. versions only needed for migration paths, not new keys.

File: Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
++            boolean isOnlineMode = server.getOnlineMode() || (SpigotConfig.bungee && PaperConfig.bungeeOnlineMode) ||
++                (PaperConfig.velocitySupport && PaperConfig.velocityOnlineMode);
+             if (isOnlineMode) {
+                 profile = lookupName ? userCache.getProfile(name) : userCache.getProfileIfCached(name);
+             } else {
+@@ -171,7 +172,8 @@ public class CraftPlayerProfile implements PlayerProfile {
+     public boolean complete(boolean textures) {
+         MinecraftServer server = MinecraftServer.getServer();
+ 
+-        boolean isOnlineMode = server.getOnlineMode() || (SpigotConfig.bungee && PaperConfig.bungeeOnlineMode);
```
Comment (aikar) [inline Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch]: We really need to abstract all this now to a single method PaperConfig.isProxyOnlineMode()

File: Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+ }
+diff --git a/src/main/java/com/destroystokyo/paper/profile/CraftPlayerProfile.java b/src/main/java/com/destroystokyo/paper/profile/CraftPlayerProfile.java
+index b151a13c1..2946722bb 100644
+--- a/src/main/java/com/destroystokyo/paper/profile/CraftPlayerProfile.java
++++ b/src/main/java/com/destroystokyo/paper/profile/CraftPlayerProfile.java
+@@ -146,7 +146,8 @@ public class CraftPlayerProfile implements PlayerProfile {
+         UserCache userCache = server.getUserCache();
+         if (profile.getId() == null) {
+             final GameProfile profile;
+-            boolean isOnlineMode = server.getOnlineMode() || (SpigotConfig.bungee && PaperConfig.bungeeOnlineMode);
```
Comment (aikar) [inline Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch]: We really need to abstract all this now to a single method PaperConfig.isProxyOnlineMode()

File: Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+         this.b = packetdataserializer.l();
+diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+index 19f2f2bbd..e723163f2 100644
+--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
++++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+@@ -1426,7 +1426,8 @@ public final class CraftServer implements Server {
+         GameProfile profile;
+         // Only fetch an online UUID in online mode
+         if (MinecraftServer.getServer().getOnlineMode()
+-                || (org.spigotmc.SpigotConfig.bungee && com.destroystokyo.paper.PaperConfig.bungeeOnlineMode)) {
```
Comment (aikar) [inline Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch]: We really need to abstract all this now to a single method PaperConfig.isProxyOnlineMode()

File: Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
++                || (org.spigotmc.SpigotConfig.bungee && com.destroystokyo.paper.PaperConfig.bungeeOnlineMode)
++                || (com.destroystokyo.paper.PaperConfig.velocitySupport && com.destroystokyo.paper.PaperConfig.velocityOnlineMode)) {
+             profile = console.getUserCache().getProfile( name );
+         } else {
+             // Make an OfflinePlayer using an offline mode UUID since the name has no profile
+@@ -1448,6 +1449,7 @@ public final class CraftServer implements Server {
+             GameProfile profile = null;
+             // Only fetch an online UUID in online mode
+             if ( MinecraftServer.getServer().getOnlineMode()
++                    || (com.destroystokyo.paper.PaperConfig.velocitySupport && com.destroystokyo.paper.PaperConfig.velocityOnlineMode) // Paper - Velocity support
```
Comment (aikar) [inline Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch]: We really need to abstract all this now to a single method PaperConfig.isProxyOnlineMode()

**Review by aikar — DISMISSED** [review]

File: Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch (hunk 4 above)
Comment (astei) [reply Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch]: I was considering that, but wondered if we wanted a separate patch for that.

Comment (aikar) [reply Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch]: We can put in the stub method in the bungee patch, then velocity amends that.

Comment (astei) [reply Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch]: Will do this tonight.

File: Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+diff --git a/src/main/java/net/minecraft/server/LoginListener.java b/src/main/java/net/minecraft/server/LoginListener.java
+index 5778a5201..d7284e3b5 100644
+--- a/src/main/java/net/minecraft/server/LoginListener.java
++++ b/src/main/java/net/minecraft/server/LoginListener.java
+@@ -43,6 +43,10 @@ public class LoginListener implements PacketLoginInListener, ITickable {
+     private SecretKey loginKey;
+     private EntityPlayer l;
+     public String hostname = ""; // CraftBukkit - add field
++    // Paper start - Velocity support
++    private static final MinecraftKey VELOCITY_CHANNEL = new MinecraftKey("velocity", "player_info");
```
Comment (kashike) [inline Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch]: `VELOCITY_PLAYER_INFO_CHANNEL`

**Review by kashike — CHANGES_REQUESTED** [review]

File: Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch [hunk 12]
```diff
@@ -0,0 +1,294 @@
...
++            section.set("enabled", false);
++            section.set("online-mode", true);
++            section.set("secret", "");
++        }
++
++        velocitySupport = getBoolean("settings.velocity-support.enabled", false);
++        velocityOnlineMode = getBoolean("settings.velocity-support.online-mode", false);
++        String secret = getString("settings.velocity-support.secret", "");
++        if (velocitySupport && secret.isEmpty()) {
++            logError("Velocity support is enabled, but no secret key was specified. Velocity support will not be enabled.");
```
Comment (electronicboy) [inline Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch]: Some part of me feels that if we're going to support this, this should be more critical

File: Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch [hunk 13]
```diff
@@ -0,0 +1,294 @@
...
++                            }
++                            // Paper end
+                             String playerName = i.getName();
+                             java.net.InetAddress address = ((java.net.InetSocketAddress) networkManager.getSocketAddress()).getAddress();
+                             java.util.UUID uniqueId = i.getId();
+@@ -325,6 +343,55 @@ public class LoginListener implements PacketLoginInListener, ITickable {
+     // Spigot end
+ 
+     public void a(PacketLoginInCustomPayload packetloginincustompayload) {
++        // Paper start - Velocity support
```
Comment (kashike) [inline Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch]: I think all the logic that can be extracted should be extracted to its own class: `io.papermc.paper.proxy.VelocityProxy` or so.

File: Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch [hunk 14]
```diff
@@ -0,0 +1,299 @@
...
+-import org.bukkit.Bukkit;
+ import org.bukkit.craftbukkit.util.Waitable;
+ import org.bukkit.event.player.AsyncPlayerPreLoginEvent;
+ import org.bukkit.event.player.PlayerPreLoginEvent;
+@@ -43,6 +41,10 @@ public class LoginListener implements PacketLoginInListener, ITickable {
+     private SecretKey loginKey;
+     private EntityPlayer l;
+     public String hostname = ""; // CraftBukkit - add field
++    // Paper start - Velocity support
++    private static final MinecraftKey VELOCITY_PLAYER_INFO_CHANNEL = new MinecraftKey("velocity", "player_info");
```
Comment (kashike) [inline Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch]: Can be moved to https://github.com/PaperMC/Paper/pull/1557/files#r223475928

File: Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch (hunk 12 above)
Comment (astei) [reply Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch]: What would be the best way to make this fatal? Perhaps `System.exit(1)`?

File: Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch (hunk 13 above)
Comment (astei) [reply Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch]: The only code I think we can extract is the logic that begins the login process in the thread pool executor, which is repeated three times in this file in this PR (and two before).

**Comment by astei** [comment]
I believe all concerns raised so far have been resolved. Should I try to merge in latest changes?

**Comment by DenWav** [comment]
Would be nice if you could rebuild patches with an updated Git client to get rid of all the superfluous changes. But that's also something we can handle at merge time.

**Comment by astei** [comment]
I've updated my Git client for the future.

**Comment by kashike** [comment]
and could you rebuild the patches with your updated client? :)

**Comment by astei** [comment]
I tried that @kashike, and it produced no measurable changes.

**Comment by astei** [comment]
After a lot of work, I managed to get the patches rebuilt so that they wouldn't touch a ton of files. Now it only touches patches that modify `PaperConfig.java`.

I'm ready for further reviews and possible merge.

File: Spigot-Server-Patches/0390-Add-Velocity-IP-Forwarding-Support.patch [hunk 15]
```diff
@@ -0,0 +1,308 @@
...
+     private static boolean metricsStarted;
++    private static boolean fatalError;
+ 
+     public static void init(File configFile) {
+         CONFIG_FILE = configFile;
+@@ -77,6 +79,11 @@ public class PaperConfig {
+         Bukkit.getLogger().severe(s);
+     }
+ 
++    protected static void fatal(String s) {
```
Comment (kashike) [inline Spigot-Server-Patches/0390-Add-Velocity-IP-Forwarding-Support.patch]: This fatal stuff should be moved to the root config patch.

**Review by kashike — CHANGES_REQUESTED** [review]

File: Spigot-Server-Patches/0390-Add-Velocity-IP-Forwarding-Support.patch [hunk 16]
```diff
@@ -0,0 +1,308 @@
...
++                this.disconnect("This server requires you to connect with Velocity.");
++                return;
++            }
++
++            if (!com.destroystokyo.paper.proxy.VelocityProxy.checkSignature(buf)) {
++                this.disconnect("Unable to verify player details");
++                return;
++            }
++
++            this.networkManager.l = new java.net.InetSocketAddress(com.destroystokyo.paper.proxy.VelocityProxy.readAddress(buf), ((java.net.InetSocketAddress) this.networkManager.getSocketAddress()).getPort());
```
Comment (zachbr) [inline Spigot-Server-Patches/0390-Add-Velocity-IP-Forwarding-Support.patch]: Add an obfuscation helper as a setter right next to the field declaration please

**Review by zachbr — CHANGES_REQUESTED** [review]

**Review by zachbr — APPROVED** [review]

File: Spigot-Server-Patches/0390-Add-Velocity-IP-Forwarding-Support.patch [hunk 17]
```diff
@@ -0,0 +1,300 @@
...
++}
+diff --git a/src/main/java/net/minecraft/server/LoginListener.java b/src/main/java/net/minecraft/server/LoginListener.java
+index 5778a5201..5feff0515 100644
+--- a/src/main/java/net/minecraft/server/LoginListener.java
++++ b/src/main/java/net/minecraft/server/LoginListener.java
+@@ -4,8 +4,7 @@ import com.destroystokyo.paper.profile.CraftPlayerProfile;
+ import com.destroystokyo.paper.profile.PlayerProfile;
+ import com.mojang.authlib.GameProfile;
+ import com.mojang.authlib.exceptions.AuthenticationUnavailableException;
+-import io.netty.channel.ChannelFuture;
```
Comment (kashike) [inline Spigot-Server-Patches/0390-Add-Velocity-IP-Forwarding-Support.patch]: Don't touch imports.

**Review by kashike — CHANGES_REQUESTED** [review]

**Comment by astei** [comment]
@kashike All import-related concerns should be resolved now. I've squashed my commits as well.

**Pull request merged**